# Proyek Akhir: Menyelesaikan Permasalahan Perusahaan Edutech

- Nama: Putri Pita Mutia
- Email: putripitamutia99@gmail.com
- Id Dicoding: pita9kk

Penjelasan mengenai kolom pada dataset Jaya Jaya Maju

- EmployeeId - Identitas karyawan
- Attrition - Apakah karyawan keluar dari perusahaan? (0=no, 1=yes)
- Age - Umur karyawan
- BusinessTravel - Komitment bepergian untuk pekerjaan
- DailyRate - Pendapatan harian (bukan gaji harian, karena ini rate pada dataset IBM)
- Department - Departemen karyawan
- DistanceFromHome - Jarak dari perusahaan ke tempat tinggal (in km)
- Education - 1 - Di bawah perguruan tinggi, 2 - Diploma, 3 - Sarjana, 4 - Magister, 5 - Doktor
- EducationField - Bidang pendidikan
- EnvironmentSatisfaction - 1-Rendah, 2-Sedang, 3-Tinggi, 4-Sangat Tinggi
- Gender - Jenis kelamin karyawan
- HourlyRate - Pendapatan per jam
- JobInvolvement - 1-Rendah, 2-Sedang, 3-Tinggi, 4-Sangat Tinggi
- JobLevel - Tingkatan pekerjaan (1 to 5)
- JobRole - Jabatan pekerjaan
- JobSatisfaction - 1-Rendah, 2-Sedang, 3-Tinggi, 4-Sangat Tinggi
- MaritalStatus - Status pernikahan
- MonthlyIncome - Pendapatan  per bulan
- MonthlyRate - Tarif bulanan
- NumCompaniesWorked - Jumlah perusahaan tempat pernah bekerja
- Over18 - Berusia di atas 18 tahun?
- OverTime - Apakah karyawan tersebut lembur?
- PercentSalaryHike - Persentase kenaikan gaji tahun terakhir
- PerformanceRating - 1 - Rendah, 2 - Baik, 3 - Sangat Baik, 4 - Sangat Baik Sekali
- RelationshipSatisfaction - 1-Rendah, 2-Sedang, 3-Tinggi, 4-Sangat Tinggi
- StandardHours - Standar jam kerja
- StockOptionLevel - Tingkat opsi saham yang dimiliki
- TotalWorkingYears - Total pengalaman kerja selama karier
- TrainingTimesLastYear - Jumlah pelatihan yang diikuti tahun lalu
- WorkLifeBalance - 1-Low, 2-Good, 3-Excellent, 4-Outstanding
- YearsAtCompany - Lama bekerja di perusahaan
- YearsInCurrentRole - Lama bekerja pada jabatan saat ini
- YearsSinceLastPromotion - Lama sejak promosi terakhir
- YearsWithCurrManager - Lama bekerja dengan manajer saat ini


## Persiapan

### Menyiapkan library yang dibutuhkan

In [1]:
import pandas as pd
import seaborn as sns
from sqlalchemy import create_engine


### Menyiapkan data yang akan digunakan


In [2]:
URL = "postgresql://postgres.vzslvkwtkzudtwvenrjr:AYOBISADONG1234@aws-0-ap-southeast-1.pooler.supabase.com:6543/postgres"

engine = create_engine(URL)

df = pd.read_sql_table("employee_data", con=engine.connect())

print(df.head())



   index  EmployeeId  Age  Attrition     BusinessTravel  DailyRate  \
0      0           1   38        NaN  Travel_Frequently       1444   
1      1           2   37        1.0      Travel_Rarely       1141   
2      2           3   51        1.0      Travel_Rarely       1323   
3      3           4   42        0.0  Travel_Frequently        555   
4      4           5   40        NaN      Travel_Rarely       1194   

               Department  DistanceFromHome  Education EducationField  ...  \
0         Human Resources                 1          4          Other  ...   
1  Research & Development                11          2        Medical  ...   
2  Research & Development                 4          4  Life Sciences  ...   
3                   Sales                26          3      Marketing  ...   
4  Research & Development                 2          4        Medical  ...   

   RelationshipSatisfaction  StandardHours StockOptionLevel  \
0                         2             80     

## Data Understanding

### Menampilkan Informasi Dataset

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 36 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     1470 non-null   int64  
 1   EmployeeId                1470 non-null   int64  
 2   Age                       1470 non-null   int64  
 3   Attrition                 1058 non-null   float64
 4   BusinessTravel            1470 non-null   object 
 5   DailyRate                 1470 non-null   int64  
 6   Department                1470 non-null   object 
 7   DistanceFromHome          1470 non-null   int64  
 8   Education                 1470 non-null   int64  
 9   EducationField            1470 non-null   object 
 10  EmployeeCount             1470 non-null   int64  
 11  EnvironmentSatisfaction   1470 non-null   int64  
 12  Gender                    1470 non-null   object 
 13  HourlyRate                1470 non-null   int64  
 14  JobInvol

**Insight:**

Berdasarkan informasi dataset yang diperoleh, dataset Jaya Jaya Maju berjumlah 1.470 baris data dan 36 atribut. Namun, terdapat perbedaan jumlah data pada kolom Attrition, pada kolom tersebut hanya terdapat **1.058 nilai non-null** dari total 1.470 data.

Perbedaan ini menunjukkan adanya missing value pada kolom Attrition. Karena Attrition merupakan variabel target yang digunakan dalam analisis, maka perlu dilakukan pemeriksaan lebih lanjut terhadap missing values tersebut sebelum memasuki tahap analisis

### Pemeriksaan Missing Value dan Duplicate

In [5]:
df.isnull().sum()

,0
index,0
EmployeeId,0
Age,0
Attrition,412
BusinessTravel,0
DailyRate,0
Department,0
DistanceFromHome,0
Education,0
EducationField,0


In [6]:
df.duplicated().sum()

np.int64(0)

**Insight:**

Dari informasi yang diperoleh, dapat dilihat bahwa kolom attrition memiliki missing value sebanyak 412 data.

Sebanyak 412 data tidak memiliki label attrition sehingga tidak dapat digunakan dalam analisis faktor penyebab attrition. Data tersebut dapat menjadi objek penelitian lanjutan menggunakan pendekatan machine learning semi-supervised untuk memperkirakan status attrition

Untuk 412 data yang tidak memiliki label attrition tidak akan digunakan dalam analisis faktor penyebab attrition yang ada nantinya menggunakan dashboard karena kolom tersebut merupakan variabel target yang menunjukkan status karyawan keluar atau tidak.



###  Statistik Deksriptif

In [7]:
df.describe()

,index,EmployeeId,Age,Attrition,DailyRate,DistanceFromHome,Education,EmployeeCount,EnvironmentSatisfaction,HourlyRate,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
count,1470.000000,1470.000000,1470.000000,1058.000000,1470.000000,1470.000000,1470.000000,1470.0,1470.000000,1470.000000,...,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000
mean,734.500000,735.500000,36.923810,0.169187,802.485714,9.192517,2.912925,1.0,2.721769,65.891156,...,2.712245,80.0,0.793878,11.279592,2.799320,2.761224,7.008163,4.229252,2.187755,4.123129
std,424.496761,424.496761,9.135373,0.375094,403.509100,8.106864,1.024165,0.0,1.093082,20.329428,...,1.081209,0.0,0.852077,7.780782,1.289271,0.706476,6.126525,3.623137,3.222430,3.568136
min,0.000000,1.000000,18.000000,0.000000,102.000000,1.000000,1.000000,1.0,1.000000,30.000000,...,1.000000,80.0,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,367.250000,368.250000,30.000000,0.000000,465.000000,2.000000,2.000000,1.0,2.000000,48.000000,...,2.000000,80.0,0.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000
50%,734.500000,735.500000,36.000000,0.000000,802.000000,7.000000,3.000000,1.0,3.000000,66.000000,...,3.000000,80.0,1.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000
75%,1101.750000,1102.750000,43.000000,0.000000,1157.000000,14.000000,4.000000,1.0,4.000000,83.750000,...,4.000000,80.0,1.000000,15.000000,3.000000,3.000000,9.000000,7.000000,3.000000,7.000000
max,1469.000000,1470.000000,60.000000,1.000000,1499.000000,29.000000,5.000000,1.0,4.000000,100.000000,...,4.000000,80.0,3.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000


In [8]:
df.describe(include="object")

,BusinessTravel,Department,EducationField,Gender,JobRole,MaritalStatus,Over18,OverTime
count,1470,1470,1470,1470,1470,1470,1470,1470
unique,3,3,6,2,9,3,1,2
top,Travel_Rarely,Research & Development,Life Sciences,Male,Sales Executive,Married,Y,No
freq,1043,961,606,882,326,673,1470,1054


**Insight:**

Berdasarkan dari informasi yang diperoleh dari statistik deksriptif, terdapat ebberapa informasi yang dapat dilihat:

- Rata-rata "Age" adalah 36-37 tahun, dengan rentang usia antara 18 hingga 60 tahun
- Rata-rata "DistanceFromHome" yakni 9,19 km, dengan jarak maksimum mencapai 29 km
- Rata-rata "TotalWorkingYears" adalah sekitar 11 tahun, sedangkan rata-rata YearsAtCompany sekitar 7 tahun.
- Rata-rata dari variabel target "Attrition yakni sebesar 0,169 yang menunjukkan bahwa sekitar 16,9% karyawan mengalami attrition, sedangkan sisanya tetap bekerja di perusahaan

### Attrition Rate Based on OverTime

In [ ]:
df.groupby("OverTime")["Attrition"].mean()

,Attrition
OverTime,
No,0.107856
Yes,0.319218


**Insight:**

Berdasarkan hasil perhitungan yang diperoleh dari rata-rata nilai Attrition pada setiap kategori OverTime, diketahui bahwa kelompok karyawan yang bekerja lembur memiliki attrition rate yang lebih tinggi dibandingkan kelompok yang tidak bekerja lembur.

### Attrition Rate Based on Departement

In [ ]:
df.groupby("Department")["Attrition"].mean()

,Attrition
Department,
Human Resources,0.157895
Research & Development,0.152639
Sales,0.206897


**Insight:**

Berdasarkan hasil perhitungan yang diperoleh dari rata-rata nilai Attrition pada setiap Department, diketahui bahwa Department Sales attrition rate tertinggi yakni 0,20. Nilai tersebut lebih tinggi dibandingkan department lainnya.

### Attrition Rate Based on JobRole

In [ ]:
df.groupby("JobRole")["Attrition"].mean()

,Attrition
JobRole,
Healthcare Representative,0.090909
Human Resources,0.200000
Laboratory Technician,0.260638
Manager,0.063291
Manufacturing Director,0.065421
Research Director,0.032258
Research Scientist,0.177570
Sales Executive,0.168103
Sales Representative,0.431034


**Insight:**

Berdasarkan hasil perhitungan yang diperoleh dari rata-rata nilai Attrition pada setiap JobRole, diketahui bahwa JobRole dari Sales Representative memiliki attrition rate tertinggi yakni 0,43. Nilai tersebut lebih tinggi dibandingkan JobRole lainnya

### Attrition Rate Based on WorkLifeBalance

In [ ]:
df.groupby("WorkLifeBalance")["Attrition"].mean()

,Attrition
WorkLifeBalance,
1,0.321429
2,0.179283
3,0.147335
4,0.194690


**Insight:**

Berdasarkan hasil perhitungan yang diperoleh dari rata-rata nilai Attrition pada setiap kategori WorkLifeBalance, diketahui bahwa kategori Low (1) memiliki attrition rate tertinggi yakni 0,32. NIlai tersebut lebih tinggi dibandingkan dengan kategori lainnya

In [ ]:
df.groupby("JobSatisfaction")["Attrition"].mean()

,Attrition
JobSatisfaction,
1,0.224390
2,0.164103
3,0.194969
4,0.114706


**Insight:**

Berdasarkan hasil perhitungan yang diperoleh dari rata-rata nilai Attrition pada setiap kategori JobSatisfaction, diketahui bahwa kategori Low (1) memiliki attrition rate tertinggi yakni 0,22. NIlai tersebut lebih tinggi dibandingkan dengan kategori lainnya

## Data Preparation / Preprocessing

Pada proses ini akan dilakukan data preparation untuk tahapan modeling dan visualisasi. Pada  tahapan modeling akan dilakukan drop kolom yakni pada kolom Attrition yang memiliki missing value pada beberapa baris. Kemudian pada kolom yang tidak memberikan informasi yang berkaitan dengan analisis yang dilakukan seperti kolom:


1.   StandardHours
2.   Over18



In [ ]:
df_clean = df.dropna(subset=["Attrition"]).copy()

In [ ]:
drop_columns = [
    "StandardHours",
    "Over18"
]

df_clean = df_clean.drop(columns=drop_columns)

In [ ]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1058 entries, 1 to 1469
Data columns (total 34 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   index                     1058 non-null   int64  
 1   EmployeeId                1058 non-null   int64  
 2   Age                       1058 non-null   int64  
 3   Attrition                 1058 non-null   float64
 4   BusinessTravel            1058 non-null   object 
 5   DailyRate                 1058 non-null   int64  
 6   Department                1058 non-null   object 
 7   DistanceFromHome          1058 non-null   int64  
 8   Education                 1058 non-null   int64  
 9   EducationField            1058 non-null   object 
 10  EmployeeCount             1058 non-null   int64  
 11  EnvironmentSatisfaction   1058 non-null   int64  
 12  Gender                    1058 non-null   object 
 13  HourlyRate                1058 non-null   int64  
 14  JobInvolvemen

**Insight:**


Setelah dilakukan drop missing value pada kolom attrition, diperoleh total dataset yakni sebanyak 1058 yang tersimpan didalam ```df_clean```. Tahapan selanjutnya dilakukan analisis dan visualisasi dengan menggunakan dataset terbaru pada Metabase